docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -40,61 +41,87 @@ Currently, the `x/auth` module provides a `DeductFeeDecorator` that deducts the
+  uint64 subspace_id = 1 [
+    (gogoproto.customname) = "SubspaceID",
+    (gogoproto.moretags) = "yaml:\"subspace_id\""
+  ];
+  // Address of the user that granted the allowance
+  string granter = 2 [ (gogoproto.moretags) = "yaml:\"granter\"" ];
+  // Address of the user being granted the allowance
+  google.protobuf.Any grantee = 3 [
+    (cosmos_proto.accepts_interface) = "Grantee",
+    (gogoproto.moretags) = "yaml:\"grantee\""
+  ];
+  // Allowance can be any allowance type implementing the FeeAllowanceI
+  // interface
+  google.protobuf.Any allowance = 4 [
+    (cosmos_proto.accepts_interface) = "FeeAllowanceI",
+    (gogoproto.moretags) = "yaml:\"allowance\""
+  ];
+}
 
-In order to represent a fee grant granted to a user group, we will implement a `GroupGrant` type. This will contain the `granter`, `group_id` and what kind of `allowance` is granted to the group:
+// UserGrantee contains the target of a grant about a user
+message UserGrantee {
+  option (gogoproto.equal) = true;
+  option (gogoproto.goproto_stringer) = true;
 
-```protobuf
-message GroupGrant {
-    // granter is the address of the user granting an allowance of their funds.
-  string granter = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string user = 1 [ (gogoproto.moretags) = "yaml:\"user\"" ];
+}
 
-  // the id of the group being granted an allowance of another user's funds.
-  uint32 group_id = 2;
+// GroupGrantee contains the target of a grant about a group
+message GroupGrantee {
+  option (gogoproto.equal) = true;
+  option (gogoproto.goproto_stringer) = true;
 
-  // allowance can be any of basic, periodic, allowed fee allowance.
-  google.protobuf.Any allowance = 3 [(cosmos_proto.accepts_interface) = "FeeAllowanceI"];
+  uint32 group_id = 1 [
+    (gogoproto.customname) = "GroupID",
+    (gogoproto.moretags) = "yaml:\"group_id\""
+  ];
 }
 ```
 
 ### Store
 
+#### User grant
+
+Each subspace fee granted allowance will be stored using keys with the following structure:
+```
+SubspaceUserAllowancePrefix | SubspaceID | GranteeAddress |-> Protobuf(Grant)
+```
+
+This structure allows subspace owner/admins to easily manage their subspace grants by iterating over all grants with the given subspace id, which will be the most used query. On the other hand, it allows the following queries:
+1. retrieving a specific grant.  
+   This will require grantees to know who the granter address, and can be done in O(1).
+2. retrieving all the grants that a user has been granted. 
+   This can be done by iterating over all the subspace's grants in O(N), where N is the number of all the grants inside the given subspace.
+
 #### Group grant
 
 In order to simplify the management of group allowances, each granted group allowance will be stored using the following key:
 ```
-SubspaceGroupAllowancePrefix | SubspaceID | GranterAddress | GroupID | -> Protobuf(GroupGrant)
+SubspaceGroupAllowancePrefix | SubspaceID | GroupID | -> Protobuf(Grant)
 ```
```

Comment:
Although `UserGrant` state can be taken care by `x/authz`, I still decided to store the state in `x/subspaces` since it is in a particular way highly depending `x/subspaces`. It also avoid the changes of `x/authz` keeper influencing `x/subspaces`.